{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:From /home/bokai/miniconda3/envs/muse/lib/python3.6/site-packages/tensorflow/python/compat/v2_compat.py:101: disable_resource_variables (from tensorflow.python.ops.variable_scope) is deprecated and will be removed in a future version.\n",
      "Instructions for updating:\n",
      "non-resource variables are not supported in the long term\n"
     ]
    }
   ],
   "source": [
    "import muse_sc as muse\n",
    "import phenograph\n",
    "from sklearn.decomposition import PCA\n",
    "import numpy as np\n",
    "from sklearn.metrics.cluster import adjusted_rand_score\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.manifold import TSNE\n",
    "import tensorflow as tf\n",
    "tf.get_logger().setLevel('ERROR')\n",
    "np.random.seed(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 245.41244220733643 seconds\n",
      "Jaccard graph constructed in 55.50883913040161 seconds\n",
      "Wrote graph to binary file in 9.867979288101196 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.858122\n",
      "After 4 runs, maximum modularity is Q = 0.860914\n",
      "Louvain completed 24 runs in 160.12310242652893 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 475.82967734336853 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 273.0466215610504 seconds\n",
      "Jaccard graph constructed in 52.369524002075195 seconds\n",
      "Wrote graph to binary file in 8.484092950820923 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.670653\n",
      "Louvain completed 21 runs in 100.2457869052887 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 440.96811962127686 seconds\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "###### process protein modality\n",
    "\n",
    "data1 = pd.read_csv(\"../data/features_and_metadata.csv\")\n",
    "features_names = [\n",
    "        'CD2', 'CD56',\n",
    "        'CD20', 'CD30', 'CD31', 'CD4', \n",
    "        'CD57',  'CD7', 'CD8',\n",
    "        'Granzyme.B','FoxP3', 'CD11c', 'CD16', 'CD206', 'CD163', 'CD68', 'CD15', 'CD11b',\n",
    "        'MCT', 'Cytokeratin', \n",
    "        'Podoplanin', 'a.SMA'\n",
    "    ]\n",
    "dataa = data1[features_names]\n",
    "#dataa = dataa.drop('nucl',axis = 1)\n",
    "#dataa = data1\n",
    "\n",
    "## change to scaled version \n",
    "from scipy.stats import zscore\n",
    "dataa = dataa.apply(zscore)\n",
    "dataa = dataa\n",
    "\n",
    "####### process imaging modality\n",
    "\n",
    "datab_full = np.load('../data/images/Inception_img_feature_chlcodex_full.npy') # full data size version\n",
    "\n",
    "###### this is pca on feature mod and image mod\n",
    "\n",
    "latent_dim = 20\n",
    "view_a_feature = PCA(n_components=latent_dim).fit_transform(dataa)\n",
    "### this is pca on img mod\n",
    "latent_dim = 100\n",
    "view_b_feature = PCA(n_components=latent_dim).fit_transform(datab_full)\n",
    "### then default clustering\n",
    "\n",
    "# already finished running\n",
    "view_a_label, _, _ = phenograph.cluster(view_a_feature, n_jobs = 16)\n",
    "view_b_label, _, _ = phenograph.cluster(view_b_feature, n_jobs = 16)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(143730, 22)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataa.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(143729,), (143729,)]\n"
     ]
    }
   ],
   "source": [
    "view_a_label = view_a_label[:-1] # last image was missing lets discard this cell\n",
    "np.save('../data/musechlcdxfull_a_label.npy', view_a_label)\n",
    "np.save('../data/musechlcdxfull_b_label.npy', view_b_label)\n",
    "print([view_a_label.shape, view_b_label.shape])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(143729,), (143729,)]\n"
     ]
    }
   ],
   "source": [
    "#view_a_label = view_a_label[:-1]\n",
    "view_a_label = np.load('../data/musechlcdxfull_a_label.npy')\n",
    "view_b_label = np.load('../data/musechlcdxfull_b_label.npy')\n",
    "print([view_a_label.shape, view_b_label.shape])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### change original muse function phenograph cpu allocation too many issue "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from muse_sc.muse_architecture import structured_embedding\n",
    "from scipy.spatial.distance import pdist\n",
    "import phenograph\n",
    "import tensorflow.compat.v1 as tf\n",
    "\n",
    "tf.disable_v2_behavior()\n",
    "\n",
    "def muse_fit_predict(data_x,\n",
    "                     data_y,\n",
    "                     label_x,\n",
    "                     label_y,\n",
    "                     latent_dim=100,\n",
    "                     n_epochs=500,\n",
    "                     lambda_regul=5,\n",
    "                     lambda_super=5):\n",
    "    \"\"\"\n",
    "        MUSE model fitting and predicting:\n",
    "          This function is used to train the MUSE model on multi-modality data\n",
    "\n",
    "        Parameters:\n",
    "          data_x:       input for transcript modality; matrix of  n * p, where n = number of cells, p = number of genes.\n",
    "          data_y:       input for morphological modality; matrix of n * q, where n = number of cells, q is the feature dimension.\n",
    "          label_x:      initial reference cluster label for transcriptional modality.\n",
    "          label_y:      inital reference cluster label for morphological modality.\n",
    "          latent_dim:   feature dimension of joint latent representation.\n",
    "          n_epochs:     maximal epoch used in training.\n",
    "          lambda_regul: weight for regularization term in the loss function.\n",
    "          lambda_super: weight for supervised learning loss in the loss function.\n",
    "\n",
    "        Output:\n",
    "          latent:       joint latent representation learned by MUSE.\n",
    "          reconstruct_x:reconstructed feature matrix corresponding to input data_x.\n",
    "          reconstruct_y:reconstructed feature matrix corresponding to input data_y.\n",
    "          latent_x:     modality-specific latent representation corresponding to data_x.\n",
    "          latent_y:     modality-specific latent representation corresponding to data_y.\n",
    "\n",
    "        Feng Bao @ Altschuler & Wu Lab @ UCSF 2022.\n",
    "        Software provided as is under MIT License.\n",
    "    \"\"\"\n",
    "\n",
    "    \"\"\" initial parameter setting \"\"\"\n",
    "    # parameter setting for neural network\n",
    "    n_hidden = 128  # number of hidden node in neural network\n",
    "    learn_rate = 1e-4  # learning rate in the optimization\n",
    "    batch_size = 64  # number of cells in the training batch\n",
    "    n_epochs_init = 200  # number of training epoch in model initialization\n",
    "    print_epochs = 50  # epoch interval to display the current training loss\n",
    "    cluster_update_epoch = 200  # epoch interval to update modality-specific clusters\n",
    "\n",
    "    # read data-specific parameters from inputs\n",
    "    feature_dim_x = data_x.shape[1]\n",
    "    feature_dim_y = data_y.shape[1]\n",
    "    n_sample = data_x.shape[0]\n",
    "\n",
    "    # GPU configuration\n",
    "    # config = tf.ConfigProto()\n",
    "    # config.gpu_options.allow_growth = True\n",
    "\n",
    "    \"\"\" construct computation graph using TensorFlow \"\"\"\n",
    "    tf.reset_default_graph()\n",
    "\n",
    "    # raw data from two modalities\n",
    "    x = tf.placeholder(tf.float32, shape=[None, feature_dim_x], name='input_x')\n",
    "    y = tf.placeholder(tf.float32, shape=[None, feature_dim_y], name='input_y')\n",
    "\n",
    "    # labels inputted for references\n",
    "    ref_label_x = tf.placeholder(tf.float32, shape=[None], name='ref_label_x')\n",
    "    ref_label_y = tf.placeholder(tf.float32, shape=[None], name='ref_label_y')\n",
    "\n",
    "    # hyperparameter in triplet loss\n",
    "    triplet_lambda = tf.placeholder(tf.float32, name='triplet_lambda')\n",
    "    triplet_margin = tf.placeholder(tf.float32, name='triplet_margin')\n",
    "\n",
    "    # network architecture\n",
    "    z, x_hat, y_hat, encode_x, encode_y, loss, \\\n",
    "    reconstruction_error, weight_penalty, \\\n",
    "    trip_loss_x, trip_loss_y = structured_embedding(x,\n",
    "                                                    y,\n",
    "                                                    ref_label_x,\n",
    "                                                    ref_label_y,\n",
    "                                                    latent_dim,\n",
    "                                                    triplet_margin,\n",
    "                                                    n_hidden,\n",
    "                                                    lambda_regul,\n",
    "                                                    triplet_lambda)\n",
    "    # optimization operator\n",
    "    train_op = tf.train.AdamOptimizer(learn_rate).minimize(loss)\n",
    "    print('++++++++++ MUSE for multi-modality single-cell analysis ++++++++++')\n",
    "    \"\"\" MUSE optimization \"\"\"\n",
    "    total_batch = int(n_sample / batch_size)\n",
    "\n",
    "    with tf.Session() as sess:\n",
    "\n",
    "        \"\"\" initialization of autoencoder architecture for MUSE \"\"\"\n",
    "        print('MUSE initialization')\n",
    "        # global parameter initialization\n",
    "        sess.run(tf.global_variables_initializer(), feed_dict={triplet_lambda: 0,\n",
    "                                                               triplet_margin: 0})\n",
    "\n",
    "        for epoch in range(n_epochs_init):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "\n",
    "            for i in range(total_batch):\n",
    "                # input data batches\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "\n",
    "                # initialize parameters without self-supervised loss (triplet_lambda=0)\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: np.zeros(batch_x_input.shape[0]),\n",
    "                                    ref_label_y: np.zeros(batch_y_input.shape[0]),\n",
    "                                    triplet_lambda: 0,\n",
    "                                    triplet_margin: 0})\n",
    "\n",
    "            # calculate and print loss terms for current epoch\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: np.zeros(data_train_x.shape[0]),  # no use as triplet_lambda=0\n",
    "                                        ref_label_y: np.zeros(data_train_y.shape[0]),  # no use as triplet_lambda=0\n",
    "                                        triplet_lambda: 0,\n",
    "                                        triplet_margin: 0})\n",
    "\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight))\n",
    "\n",
    "        # estimate the margin for the triplet loss\n",
    "        latent, reconstruct_x, reconstruct_y = \\\n",
    "            sess.run((z, x_hat, y_hat),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: np.zeros(data_x.shape[0]),\n",
    "                                ref_label_y: np.zeros(data_y.shape[0]),\n",
    "                                triplet_lambda: 0,\n",
    "                                triplet_margin: 0})\n",
    "        latent_pd_matrix = pdist(latent, 'euclidean')\n",
    "        latent_pd_sort = np.sort(latent_pd_matrix)\n",
    "        select_top_n = np.int(latent_pd_sort.size * 0.2)\n",
    "        margin_estimate = np.median(latent_pd_sort[-select_top_n:]) - np.median(latent_pd_sort[:select_top_n])\n",
    "\n",
    "        # refine MUSE parameters with reference labels and triplet losses\n",
    "        for epoch in range(n_epochs_init):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "            label_train_x = label_x[random_idx]\n",
    "            label_train_y = label_y[random_idx]\n",
    "\n",
    "            for i in range(total_batch):\n",
    "                # data batches\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "                label_x_input = label_train_x[offset:(offset + batch_size)]\n",
    "                label_y_input = label_train_y[offset:(offset + batch_size)]\n",
    "\n",
    "                # refine parameters\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: label_x_input,\n",
    "                                    ref_label_y: label_y_input,\n",
    "                                    triplet_lambda: lambda_super,\n",
    "                                    triplet_margin: margin_estimate})\n",
    "\n",
    "            # calculate loss on all input data for current epoch\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty, trip_loss_x, trip_loss_y),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: label_train_x,\n",
    "                                        ref_label_y: label_train_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f,\\t x triplet: %03.5f,\\t y triplet: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y))\n",
    "\n",
    "        # update cluster labels based modality-specific latents\n",
    "        latent_x, latent_y = \\\n",
    "            sess.run((encode_x, encode_y),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: label_x,\n",
    "                                ref_label_y: label_y,\n",
    "                                triplet_lambda: lambda_super,\n",
    "                                triplet_margin: margin_estimate})\n",
    "\n",
    "        # update cluster labels using PhenoGraph\n",
    "        label_x_update, _, _ = phenograph.cluster(latent_x, n_jobs= 2)\n",
    "        label_y_update, _, _ = phenograph.cluster(latent_y, n_jobs= 2)\n",
    "        print('Finish initialization of MUSE')\n",
    "\n",
    "        ''' Training of MUSE '''\n",
    "        for epoch in range(n_epochs):\n",
    "            # randomly permute samples\n",
    "            random_idx = np.random.permutation(n_sample)\n",
    "            data_train_x = data_x[random_idx, :]\n",
    "            data_train_y = data_y[random_idx, :]\n",
    "            label_train_x = label_x_update[random_idx]\n",
    "            label_train_y = label_y_update[random_idx]\n",
    "\n",
    "            # loop over all batches\n",
    "            for i in range(total_batch):\n",
    "                # batch data\n",
    "                offset = (i * batch_size) % (n_sample)\n",
    "                batch_x_input = data_train_x[offset:(offset + batch_size), :]\n",
    "                batch_y_input = data_train_y[offset:(offset + batch_size), :]\n",
    "                batch_label_x_input = label_train_x[offset:(offset + batch_size)]\n",
    "                batch_label_y_input = label_train_y[offset:(offset + batch_size)]\n",
    "\n",
    "                sess.run(train_op,\n",
    "                         feed_dict={x: batch_x_input,\n",
    "                                    y: batch_y_input,\n",
    "                                    ref_label_x: batch_label_x_input,\n",
    "                                    ref_label_y: batch_label_y_input,\n",
    "                                    triplet_lambda: lambda_super,\n",
    "                                    triplet_margin: margin_estimate})\n",
    "\n",
    "            # calculate and print losses on whole training dataset\n",
    "            if epoch % print_epochs == 0:\n",
    "                L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y = \\\n",
    "                    sess.run((loss, reconstruction_error, weight_penalty, trip_loss_x, trip_loss_y),\n",
    "                             feed_dict={x: data_train_x,\n",
    "                                        y: data_train_y,\n",
    "                                        ref_label_x: label_train_x,\n",
    "                                        ref_label_y: label_train_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "                # print cost every epoch\n",
    "                print(\n",
    "                    \"epoch: %d, \\t total loss: %03.5f,\\t reconstruction loss: %03.5f,\\t sparse penalty: %03.5f,\\t x triplet loss: %03.5f,\\t y triplet loss: %03.5f\"\n",
    "                    % (epoch, L_total, L_reconstruction, L_weight, L_trip_x, L_trip_y))\n",
    "\n",
    "            # update cluster labels based on new modality-specific latent representations\n",
    "            if epoch % cluster_update_epoch == 0:\n",
    "                latent_x, latent_y = \\\n",
    "                    sess.run((encode_x, encode_y),\n",
    "                             feed_dict={x: data_x,\n",
    "                                        y: data_y,\n",
    "                                        ref_label_x: label_x,\n",
    "                                        ref_label_y: label_y,\n",
    "                                        triplet_lambda: lambda_super,\n",
    "                                        triplet_margin: margin_estimate})\n",
    "\n",
    "                # use PhenoGraph to obtain cluster label\n",
    "                label_x_update, _, _ = phenograph.cluster(latent_x, n_jobs= 2)\n",
    "                label_y_update, _, _ = phenograph.cluster(latent_y, n_jobs= 2)\n",
    "\n",
    "        \"\"\" MUSE output \"\"\"\n",
    "        latent, reconstruct_x, reconstruct_y, latent_x, latent_y = \\\n",
    "            sess.run((z, x_hat, y_hat, encode_x, encode_y),\n",
    "                     feed_dict={x: data_x,\n",
    "                                y: data_y,\n",
    "                                ref_label_x: label_x,  # no effects to representations\n",
    "                                ref_label_y: label_y,  # no effects to representations\n",
    "                                triplet_lambda: lambda_super,\n",
    "                                triplet_margin: margin_estimate})\n",
    "\n",
    "        print('++++++++++ MUSE completed ++++++++++')\n",
    "\n",
    "    return latent, reconstruct_x, reconstruct_y, latent_x, latent_y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1523.49414,\t reconstruction loss: 1229.62915,\t sparse penalty: 58.77299\n",
      "epoch: 50, \t total loss: 841.77588,\t reconstruction loss: 694.78754,\t sparse penalty: 29.39767\n",
      "epoch: 100, \t total loss: 705.31390,\t reconstruction loss: 647.00122,\t sparse penalty: 11.66253\n",
      "epoch: 150, \t total loss: 647.97119,\t reconstruction loss: 633.75745,\t sparse penalty: 2.84275\n",
      "epoch: 0, \t total loss: 802.07434,\t reconstruction loss: 691.57196,\t sparse penalty: 0.41859,\t x triplet: 11.05236,\t y triplet: 10.62953\n",
      "epoch: 50, \t total loss: 718.78711,\t reconstruction loss: 636.20740,\t sparse penalty: 0.55731,\t x triplet: 8.03356,\t y triplet: 7.92508\n",
      "epoch: 100, \t total loss: 711.66370,\t reconstruction loss: 632.14392,\t sparse penalty: 0.41956,\t x triplet: 7.78116,\t y triplet: 7.70323\n",
      "epoch: 150, \t total loss: 707.21643,\t reconstruction loss: 629.25079,\t sparse penalty: 0.34147,\t x triplet: 7.65914,\t y triplet: 7.59253\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.5714406967163086 seconds\n",
      "Jaccard graph constructed in 10.931543827056885 seconds\n",
      "Wrote graph to binary file in 0.750990629196167 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.85342\n",
      "After 2 runs, maximum modularity is Q = 0.85494\n",
      "Louvain completed 22 runs in 8.965057373046875 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 30.777803659439087 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.117755889892578 seconds\n",
      "Jaccard graph constructed in 11.040194034576416 seconds\n",
      "Wrote graph to binary file in 0.9114723205566406 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.598232\n",
      "Louvain completed 21 runs in 9.225244760513306 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 32.29442286491394 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 704.84253,\t reconstruction loss: 627.77319,\t sparse penalty: 0.29551,\t x triplet loss: 7.57885,\t y triplet loss: 7.53951\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.2601959705352783 seconds\n",
      "Jaccard graph constructed in 11.125582933425903 seconds\n",
      "Wrote graph to binary file in 0.8777580261230469 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.852837\n",
      "After 5 runs, maximum modularity is Q = 0.854027\n",
      "Louvain completed 25 runs in 10.092222929000854 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 33.41950297355652 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.8225228786468506 seconds\n",
      "Jaccard graph constructed in 11.199826717376709 seconds\n",
      "Wrote graph to binary file in 0.8471379280090332 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.589298\n",
      "After 2 runs, maximum modularity is Q = 0.593185\n",
      "After 3 runs, maximum modularity is Q = 0.599117\n",
      "Louvain completed 23 runs in 11.81606125831604 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 34.97452354431152 seconds\n",
      "epoch: 50, \t total loss: 703.04456,\t reconstruction loss: 626.75214,\t sparse penalty: 0.26943,\t x triplet loss: 7.51632,\t y triplet loss: 7.47273\n",
      "epoch: 100, \t total loss: 701.77124,\t reconstruction loss: 625.94550,\t sparse penalty: 0.24594,\t x triplet loss: 7.48017,\t y triplet loss: 7.43904\n",
      "epoch: 150, \t total loss: 701.08020,\t reconstruction loss: 625.58173,\t sparse penalty: 0.22735,\t x triplet loss: 7.45514,\t y triplet loss: 7.41720\n",
      "epoch: 200, \t total loss: 700.52936,\t reconstruction loss: 625.30273,\t sparse penalty: 0.21246,\t x triplet loss: 7.43348,\t y triplet loss: 7.39938\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.2760581970214844 seconds\n",
      "Jaccard graph constructed in 11.534981489181519 seconds\n",
      "Wrote graph to binary file in 0.8961656093597412 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.849916\n",
      "Louvain completed 21 runs in 8.143360614776611 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 32.450249433517456 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.051915168762207 seconds\n",
      "Jaccard graph constructed in 12.156038999557495 seconds\n",
      "Wrote graph to binary file in 0.8332931995391846 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.59792\n",
      "Louvain completed 21 runs in 9.333055973052979 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 34.17601919174194 seconds\n",
      "epoch: 250, \t total loss: 699.29126,\t reconstruction loss: 624.33893,\t sparse penalty: 0.20711,\t x triplet loss: 7.40239,\t y triplet loss: 7.38097\n",
      "epoch: 300, \t total loss: 698.06165,\t reconstruction loss: 623.30493,\t sparse penalty: 0.19657,\t x triplet loss: 7.38811,\t y triplet loss: 7.36667\n",
      "epoch: 350, \t total loss: 698.05408,\t reconstruction loss: 623.46503,\t sparse penalty: 0.18892,\t x triplet loss: 7.37489,\t y triplet loss: 7.35400\n",
      "epoch: 400, \t total loss: 697.44678,\t reconstruction loss: 623.00183,\t sparse penalty: 0.18105,\t x triplet loss: 7.36416,\t y triplet loss: 7.34378\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.080127239227295 seconds\n",
      "Jaccard graph constructed in 12.447417259216309 seconds\n",
      "Wrote graph to binary file in 1.029567003250122 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.841874\n",
      "After 12 runs, maximum modularity is Q = 0.843202\n",
      "Louvain completed 32 runs in 12.908492803573608 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 38.434365034103394 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.1414828300476074 seconds\n",
      "Jaccard graph constructed in 12.616402626037598 seconds\n",
      "Wrote graph to binary file in 0.7607550621032715 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.599612\n",
      "After 3 runs, maximum modularity is Q = 0.601973\n",
      "After 9 runs, maximum modularity is Q = 0.604055\n",
      "Louvain completed 29 runs in 14.943246364593506 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 40.68928027153015 seconds\n",
      "epoch: 450, \t total loss: 696.22382,\t reconstruction loss: 621.90393,\t sparse penalty: 0.17452,\t x triplet loss: 7.35531,\t y triplet loss: 7.33415\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1483.48792,\t reconstruction loss: 1180.44189,\t sparse penalty: 60.60920\n",
      "epoch: 50, \t total loss: 860.38245,\t reconstruction loss: 702.09943,\t sparse penalty: 31.65660\n",
      "epoch: 100, \t total loss: 717.92749,\t reconstruction loss: 647.25244,\t sparse penalty: 14.13501\n",
      "epoch: 150, \t total loss: 664.72260,\t reconstruction loss: 636.39618,\t sparse penalty: 5.66528\n",
      "epoch: 0, \t total loss: 807.69690,\t reconstruction loss: 697.24921,\t sparse penalty: 1.94476,\t x triplet: 10.28241,\t y triplet: 9.86236\n",
      "epoch: 50, \t total loss: 718.65125,\t reconstruction loss: 641.27167,\t sparse penalty: 0.77639,\t x triplet: 7.38731,\t y triplet: 7.31221\n",
      "epoch: 100, \t total loss: 711.20398,\t reconstruction loss: 638.26947,\t sparse penalty: 0.42472,\t x triplet: 7.10464,\t y triplet: 7.05755\n",
      "epoch: 150, \t total loss: 707.91315,\t reconstruction loss: 636.62952,\t sparse penalty: 0.33969,\t x triplet: 6.97921,\t y triplet: 6.93781\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9758758544921875 seconds\n",
      "Jaccard graph constructed in 13.42635202407837 seconds\n",
      "Wrote graph to binary file in 0.8646225929260254 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.855997\n",
      "After 10 runs, maximum modularity is Q = 0.857343\n",
      "Louvain completed 30 runs in 12.759981870651245 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 39.897849321365356 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.8068718910217285 seconds\n",
      "Jaccard graph constructed in 13.80121660232544 seconds\n",
      "Wrote graph to binary file in 1.0081920623779297 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.586869\n",
      "After 2 runs, maximum modularity is Q = 0.590486\n",
      "After 9 runs, maximum modularity is Q = 0.592571\n",
      "After 26 runs, maximum modularity is Q = 0.594567\n",
      "Louvain completed 46 runs in 23.982773303985596 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 51.70225191116333 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 705.80750,\t reconstruction loss: 635.43945,\t sparse penalty: 0.29391,\t x triplet loss: 6.89605,\t y triplet loss: 6.88364\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.1798479557037354 seconds\n",
      "Jaccard graph constructed in 13.863373756408691 seconds\n",
      "Wrote graph to binary file in 0.9446170330047607 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.854986\n",
      "After 2 runs, maximum modularity is Q = 0.857191\n",
      "Louvain completed 22 runs in 10.337036371231079 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 38.721848487854004 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.116903066635132 seconds\n",
      "Jaccard graph constructed in 13.812784910202026 seconds\n",
      "Wrote graph to binary file in 0.9233677387237549 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.596382\n",
      "After 8 runs, maximum modularity is Q = 0.597447\n",
      "After 11 runs, maximum modularity is Q = 0.601087\n",
      "Louvain completed 31 runs in 16.635825634002686 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 44.93446183204651 seconds\n",
      "epoch: 50, \t total loss: 703.73108,\t reconstruction loss: 634.24988,\t sparse penalty: 0.27510,\t x triplet loss: 6.81918,\t y triplet loss: 6.80196\n",
      "epoch: 100, \t total loss: 702.79559,\t reconstruction loss: 633.80890,\t sparse penalty: 0.24983,\t x triplet loss: 6.78202,\t y triplet loss: 6.76549\n",
      "epoch: 150, \t total loss: 702.44305,\t reconstruction loss: 633.82166,\t sparse penalty: 0.23023,\t x triplet loss: 6.75326,\t y triplet loss: 6.74078\n",
      "epoch: 200, \t total loss: 701.65314,\t reconstruction loss: 633.32446,\t sparse penalty: 0.21411,\t x triplet loss: 6.73115,\t y triplet loss: 6.72047\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.809539556503296 seconds\n",
      "Jaccard graph constructed in 14.465286254882812 seconds\n",
      "Wrote graph to binary file in 0.9682431221008301 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.853076\n",
      "After 10 runs, maximum modularity is Q = 0.855416\n",
      "Louvain completed 30 runs in 13.149543762207031 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 42.14104413986206 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.1102898120880127 seconds\n",
      "Jaccard graph constructed in 14.6972074508667 seconds\n",
      "Wrote graph to binary file in 0.9294533729553223 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.600786\n",
      "After 15 runs, maximum modularity is Q = 0.607694\n",
      "Louvain completed 35 runs in 17.045142889022827 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 46.80776405334473 seconds\n",
      "epoch: 250, \t total loss: 701.60889,\t reconstruction loss: 633.52222,\t sparse penalty: 0.20232,\t x triplet loss: 6.71179,\t y triplet loss: 6.70322\n",
      "epoch: 300, \t total loss: 699.45538,\t reconstruction loss: 631.57025,\t sparse penalty: 0.19190,\t x triplet loss: 6.69636,\t y triplet loss: 6.68877\n",
      "epoch: 350, \t total loss: 698.64832,\t reconstruction loss: 630.92316,\t sparse penalty: 0.18295,\t x triplet loss: 6.68450,\t y triplet loss: 6.67757\n",
      "epoch: 400, \t total loss: 698.73590,\t reconstruction loss: 631.16919,\t sparse penalty: 0.17558,\t x triplet loss: 6.67305,\t y triplet loss: 6.66473\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.309792995452881 seconds\n",
      "Jaccard graph constructed in 15.142863988876343 seconds\n",
      "Wrote graph to binary file in 1.1544413566589355 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.842905\n",
      "After 10 runs, maximum modularity is Q = 0.844343\n",
      "Louvain completed 30 runs in 13.86020803451538 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 44.5498685836792 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.003575086593628 seconds\n",
      "Jaccard graph constructed in 15.171953678131104 seconds\n",
      "Wrote graph to binary file in 0.8528168201446533 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.601496\n",
      "After 3 runs, maximum modularity is Q = 0.60774\n",
      "Louvain completed 23 runs in 12.684877872467041 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 44.48989391326904 seconds\n",
      "epoch: 450, \t total loss: 697.86096,\t reconstruction loss: 630.44019,\t sparse penalty: 0.17056,\t x triplet loss: 6.66130,\t y triplet loss: 6.65228\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1573.69641,\t reconstruction loss: 1286.11255,\t sparse penalty: 57.51677\n",
      "epoch: 50, \t total loss: 845.70435,\t reconstruction loss: 703.61456,\t sparse penalty: 28.41795\n",
      "epoch: 100, \t total loss: 699.17731,\t reconstruction loss: 644.09296,\t sparse penalty: 11.01687\n",
      "epoch: 150, \t total loss: 648.39319,\t reconstruction loss: 633.84009,\t sparse penalty: 2.91062\n",
      "epoch: 0, \t total loss: 791.75610,\t reconstruction loss: 691.88092,\t sparse penalty: 0.63539,\t x triplet: 9.87581,\t y triplet: 9.46383\n",
      "epoch: 50, \t total loss: 710.18964,\t reconstruction loss: 637.39941,\t sparse penalty: 0.56853,\t x triplet: 7.04385,\t y triplet: 6.94565\n",
      "epoch: 100, \t total loss: 703.85400,\t reconstruction loss: 634.40265,\t sparse penalty: 0.39474,\t x triplet: 6.78083,\t y triplet: 6.71468\n",
      "epoch: 150, \t total loss: 700.22388,\t reconstruction loss: 632.42670,\t sparse penalty: 0.31375,\t x triplet: 6.64866,\t y triplet: 6.59703\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9442484378814697 seconds\n",
      "Jaccard graph constructed in 16.919131994247437 seconds\n",
      "Wrote graph to binary file in 0.9174096584320068 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.851996\n",
      "After 9 runs, maximum modularity is Q = 0.853009\n",
      "Louvain completed 29 runs in 13.812788248062134 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 47.916282176971436 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9683704376220703 seconds\n",
      "Jaccard graph constructed in 16.489983320236206 seconds\n",
      "Wrote graph to binary file in 1.1344375610351562 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.584099\n",
      "After 2 runs, maximum modularity is Q = 0.586707\n",
      "After 3 runs, maximum modularity is Q = 0.590906\n",
      "After 6 runs, maximum modularity is Q = 0.593482\n",
      "Louvain completed 26 runs in 17.379469871520996 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 51.68450665473938 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 698.54163,\t reconstruction loss: 631.62561,\t sparse penalty: 0.27161,\t x triplet loss: 6.56608,\t y triplet loss: 6.54551\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.5577175617218018 seconds\n",
      "Jaccard graph constructed in 17.191354036331177 seconds\n",
      "Wrote graph to binary file in 1.0492985248565674 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.851778\n",
      "After 5 runs, maximum modularity is Q = 0.853024\n",
      "Louvain completed 25 runs in 12.462869882583618 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 47.52842998504639 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.0547478199005127 seconds\n",
      "Jaccard graph constructed in 17.968353033065796 seconds\n",
      "Wrote graph to binary file in 1.0255208015441895 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.583948\n",
      "After 2 runs, maximum modularity is Q = 0.597113\n",
      "Louvain completed 22 runs in 12.974189043045044 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 49.240800619125366 seconds\n",
      "epoch: 50, \t total loss: 696.49078,\t reconstruction loss: 630.35724,\t sparse penalty: 0.24893,\t x triplet loss: 6.50309,\t y triplet loss: 6.47469\n",
      "epoch: 100, \t total loss: 695.98621,\t reconstruction loss: 630.29950,\t sparse penalty: 0.22639,\t x triplet loss: 6.46891,\t y triplet loss: 6.44204\n",
      "epoch: 150, \t total loss: 694.84955,\t reconstruction loss: 629.52625,\t sparse penalty: 0.20853,\t x triplet loss: 6.44166,\t y triplet loss: 6.41446\n",
      "epoch: 200, \t total loss: 693.36359,\t reconstruction loss: 628.29950,\t sparse penalty: 0.19559,\t x triplet loss: 6.42180,\t y triplet loss: 6.39542\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.8934121131896973 seconds\n",
      "Jaccard graph constructed in 18.35091280937195 seconds\n",
      "Wrote graph to binary file in 1.189012050628662 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.840774\n",
      "After 2 runs, maximum modularity is Q = 0.843029\n",
      "Louvain completed 22 runs in 11.482810735702515 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 48.548171281814575 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.650810718536377 seconds\n",
      "Jaccard graph constructed in 18.437719106674194 seconds\n",
      "Wrote graph to binary file in 1.0500280857086182 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.587704\n",
      "After 2 runs, maximum modularity is Q = 0.600495\n",
      "Louvain completed 22 runs in 13.320550680160522 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 50.4498872756958 seconds\n",
      "epoch: 250, \t total loss: 692.83710,\t reconstruction loss: 628.00220,\t sparse penalty: 0.18733,\t x triplet loss: 6.39944,\t y triplet loss: 6.38021\n",
      "epoch: 300, \t total loss: 692.02014,\t reconstruction loss: 627.36847,\t sparse penalty: 0.17829,\t x triplet loss: 6.38452,\t y triplet loss: 6.36752\n",
      "epoch: 350, \t total loss: 691.51038,\t reconstruction loss: 627.02075,\t sparse penalty: 0.16958,\t x triplet loss: 6.37264,\t y triplet loss: 6.35570\n",
      "epoch: 400, \t total loss: 690.40100,\t reconstruction loss: 626.04254,\t sparse penalty: 0.16253,\t x triplet loss: 6.36296,\t y triplet loss: 6.34619\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.8459908962249756 seconds\n",
      "Jaccard graph constructed in 19.002406120300293 seconds\n",
      "Wrote graph to binary file in 1.1095795631408691 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.834358\n",
      "After 4 runs, maximum modularity is Q = 0.835439\n",
      "Louvain completed 24 runs in 12.862164497375488 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 50.88849878311157 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9526269435882568 seconds\n",
      "Jaccard graph constructed in 19.027119159698486 seconds\n",
      "Wrote graph to binary file in 0.902019739151001 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.599316\n",
      "After 2 runs, maximum modularity is Q = 0.602648\n",
      "Louvain completed 22 runs in 13.441787004470825 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 52.042486906051636 seconds\n",
      "epoch: 450, \t total loss: 689.63818,\t reconstruction loss: 625.41699,\t sparse penalty: 0.15703,\t x triplet loss: 6.35114,\t y triplet loss: 6.33607\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1606.09338,\t reconstruction loss: 1312.95593,\t sparse penalty: 58.62749\n",
      "epoch: 50, \t total loss: 842.57745,\t reconstruction loss: 695.97327,\t sparse penalty: 29.32084\n",
      "epoch: 100, \t total loss: 702.64392,\t reconstruction loss: 645.67090,\t sparse penalty: 11.39460\n",
      "epoch: 150, \t total loss: 648.25824,\t reconstruction loss: 635.06744,\t sparse penalty: 2.63816\n",
      "epoch: 0, \t total loss: 828.99902,\t reconstruction loss: 706.76984,\t sparse penalty: 0.39821,\t x triplet: 12.27159,\t y triplet: 11.77606\n",
      "epoch: 50, \t total loss: 739.17108,\t reconstruction loss: 646.34583,\t sparse penalty: 0.56023,\t x triplet: 9.03799,\t y triplet: 8.96683\n",
      "epoch: 100, \t total loss: 728.90759,\t reconstruction loss: 638.82251,\t sparse penalty: 0.46097,\t x triplet: 8.80785,\t y triplet: 8.74819\n",
      "epoch: 150, \t total loss: 724.14014,\t reconstruction loss: 635.46613,\t sparse penalty: 0.39521,\t x triplet: 8.69480,\t y triplet: 8.64480\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.652343511581421 seconds\n",
      "Jaccard graph constructed in 19.76272940635681 seconds\n",
      "Wrote graph to binary file in 1.1718354225158691 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.856158\n",
      "After 3 runs, maximum modularity is Q = 0.857707\n",
      "Louvain completed 23 runs in 12.905921220779419 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 52.793827295303345 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 2.0187225341796875 seconds\n",
      "Jaccard graph constructed in 20.27552604675293 seconds\n",
      "Wrote graph to binary file in 1.0441274642944336 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.614035\n",
      "After 3 runs, maximum modularity is Q = 0.617016\n",
      "Louvain completed 23 runs in 13.589393377304077 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 54.11655139923096 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 722.61365,\t reconstruction loss: 634.74023,\t sparse penalty: 0.34721,\t x triplet loss: 8.62587,\t y triplet loss: 8.60161\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.77207350730896 seconds\n",
      "Jaccard graph constructed in 20.163245677947998 seconds\n",
      "Wrote graph to binary file in 0.9287011623382568 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.855671\n",
      "After 16 runs, maximum modularity is Q = 0.857037\n",
      "Louvain completed 36 runs in 18.247780084609985 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 58.63438534736633 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.995678424835205 seconds\n",
      "Jaccard graph constructed in 20.385604858398438 seconds\n",
      "Wrote graph to binary file in 1.0516507625579834 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.616496\n",
      "After 4 runs, maximum modularity is Q = 0.618396\n",
      "After 18 runs, maximum modularity is Q = 0.619846\n",
      "Louvain completed 38 runs in 22.3458149433136 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 63.71144342422485 seconds\n",
      "epoch: 50, \t total loss: 720.81671,\t reconstruction loss: 633.82330,\t sparse penalty: 0.32217,\t x triplet loss: 8.54693,\t y triplet loss: 8.52959\n",
      "epoch: 100, \t total loss: 719.14532,\t reconstruction loss: 632.67163,\t sparse penalty: 0.29494,\t x triplet loss: 8.50803,\t y triplet loss: 8.49178\n",
      "epoch: 150, \t total loss: 718.24255,\t reconstruction loss: 632.13770,\t sparse penalty: 0.27190,\t x triplet loss: 8.48273,\t y triplet loss: 8.46635\n",
      "epoch: 200, \t total loss: 716.43335,\t reconstruction loss: 630.59247,\t sparse penalty: 0.25308,\t x triplet loss: 8.46595,\t y triplet loss: 8.44915\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.6749804019927979 seconds\n",
      "Jaccard graph constructed in 20.69106364250183 seconds\n",
      "Wrote graph to binary file in 1.065345287322998 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.850964\n",
      "After 2 runs, maximum modularity is Q = 0.852141\n",
      "Louvain completed 22 runs in 12.944136142730713 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 54.625481843948364 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9952661991119385 seconds\n",
      "Jaccard graph constructed in 21.586841106414795 seconds\n",
      "Wrote graph to binary file in 1.0368258953094482 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.632378\n",
      "After 6 runs, maximum modularity is Q = 0.634578\n",
      "After 15 runs, maximum modularity is Q = 0.637574\n",
      "Louvain completed 35 runs in 21.126625061035156 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 64.21801114082336 seconds\n",
      "epoch: 250, \t total loss: 715.52026,\t reconstruction loss: 629.91479,\t sparse penalty: 0.23588,\t x triplet loss: 8.45146,\t y triplet loss: 8.43375\n",
      "epoch: 300, \t total loss: 714.45105,\t reconstruction loss: 629.03052,\t sparse penalty: 0.22315,\t x triplet loss: 8.43967,\t y triplet loss: 8.42128\n",
      "epoch: 350, \t total loss: 713.81311,\t reconstruction loss: 628.55701,\t sparse penalty: 0.21294,\t x triplet loss: 8.42990,\t y triplet loss: 8.40839\n",
      "epoch: 400, \t total loss: 712.31128,\t reconstruction loss: 627.21167,\t sparse penalty: 0.20409,\t x triplet loss: 8.41682,\t y triplet loss: 8.39901\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.8597643375396729 seconds\n",
      "Jaccard graph constructed in 21.128028392791748 seconds\n",
      "Wrote graph to binary file in 1.1586651802062988 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.846011\n",
      "After 2 runs, maximum modularity is Q = 0.847184\n",
      "After 3 runs, maximum modularity is Q = 0.848501\n",
      "After 22 runs, maximum modularity is Q = 0.849944\n",
      "Louvain completed 42 runs in 24.956156015396118 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 67.80096435546875 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.6838841438293457 seconds\n",
      "Jaccard graph constructed in 21.512765884399414 seconds\n",
      "Wrote graph to binary file in 1.040437936782837 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.632045\n",
      "After 2 runs, maximum modularity is Q = 0.635215\n",
      "Louvain completed 22 runs in 13.594372749328613 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 56.8606493473053 seconds\n",
      "epoch: 450, \t total loss: 711.50800,\t reconstruction loss: 626.53149,\t sparse penalty: 0.19681,\t x triplet loss: 8.40501,\t y triplet loss: 8.39348\n",
      "++++++++++ MUSE completed ++++++++++\n",
      "++++++++++ MUSE for multi-modality single-cell analysis ++++++++++\n",
      "MUSE initialization\n",
      "epoch: 0, \t total loss: 1505.95837,\t reconstruction loss: 1206.10010,\t sparse penalty: 59.97166\n",
      "epoch: 50, \t total loss: 849.59875,\t reconstruction loss: 696.66925,\t sparse penalty: 30.58590\n",
      "epoch: 100, \t total loss: 704.05798,\t reconstruction loss: 642.16956,\t sparse penalty: 12.37769\n",
      "epoch: 150, \t total loss: 653.64343,\t reconstruction loss: 636.21930,\t sparse penalty: 3.48482\n",
      "epoch: 0, \t total loss: 819.89404,\t reconstruction loss: 705.98853,\t sparse penalty: 1.02587,\t x triplet: 11.08246,\t y triplet: 10.67278\n",
      "epoch: 50, \t total loss: 721.77490,\t reconstruction loss: 638.32819,\t sparse penalty: 0.53470,\t x triplet: 8.12541,\t y triplet: 8.02924\n",
      "epoch: 100, \t total loss: 715.17731,\t reconstruction loss: 634.60352,\t sparse penalty: 0.40975,\t x triplet: 7.88570,\t y triplet: 7.81930\n",
      "epoch: 150, \t total loss: 712.52002,\t reconstruction loss: 633.33099,\t sparse penalty: 0.33936,\t x triplet: 7.77704,\t y triplet: 7.72142\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.871521234512329 seconds\n",
      "Jaccard graph constructed in 22.450358390808105 seconds\n",
      "Wrote graph to binary file in 1.3596415519714355 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.861509\n",
      "After 3 runs, maximum modularity is Q = 0.863045\n",
      "Louvain completed 23 runs in 13.435770750045776 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 58.82409334182739 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.7841660976409912 seconds\n",
      "Jaccard graph constructed in 22.988632440567017 seconds\n",
      "Wrote graph to binary file in 0.9957499504089355 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.595291\n",
      "After 2 runs, maximum modularity is Q = 0.59752\n",
      "Louvain completed 22 runs in 14.579777002334595 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 60.192333936691284 seconds\n",
      "Finish initialization of MUSE\n",
      "epoch: 0, \t total loss: 710.06354,\t reconstruction loss: 631.71185,\t sparse penalty: 0.29834,\t x triplet loss: 7.70236,\t y triplet loss: 7.66963\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.7735283374786377 seconds\n",
      "Jaccard graph constructed in 23.782610177993774 seconds\n",
      "Wrote graph to binary file in 1.2412664890289307 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.8629\n",
      "Louvain completed 21 runs in 11.827831506729126 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 60.136874198913574 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 14.615305662155151 seconds\n",
      "Jaccard graph constructed in 24.610965728759766 seconds\n",
      "Wrote graph to binary file in 1.430281400680542 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.590862\n",
      "After 3 runs, maximum modularity is Q = 0.599115\n",
      "After 8 runs, maximum modularity is Q = 0.600256\n",
      "After 13 runs, maximum modularity is Q = 0.6026\n",
      "Louvain completed 33 runs in 23.279808282852173 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 84.77999806404114 seconds\n",
      "epoch: 50, \t total loss: 708.70990,\t reconstruction loss: 631.13123,\t sparse penalty: 0.28178,\t x triplet loss: 7.62906,\t y triplet loss: 7.60489\n",
      "epoch: 100, \t total loss: 706.86859,\t reconstruction loss: 629.74835,\t sparse penalty: 0.25857,\t x triplet loss: 7.59411,\t y triplet loss: 7.57138\n",
      "epoch: 150, \t total loss: 706.21167,\t reconstruction loss: 629.43024,\t sparse penalty: 0.23961,\t x triplet loss: 7.56812,\t y triplet loss: 7.54856\n",
      "epoch: 200, \t total loss: 705.50378,\t reconstruction loss: 628.98663,\t sparse penalty: 0.22400,\t x triplet loss: 7.54939,\t y triplet loss: 7.53004\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 1.9460878372192383 seconds\n",
      "Jaccard graph constructed in 25.355312824249268 seconds\n",
      "Wrote graph to binary file in 1.0544145107269287 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.853203\n",
      "After 7 runs, maximum modularity is Q = 0.854254\n",
      "Louvain completed 27 runs in 17.616756200790405 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 70.68063163757324 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 39.68047761917114 seconds\n",
      "Jaccard graph constructed in 27.042834758758545 seconds\n",
      "Wrote graph to binary file in 1.039799451828003 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.599595\n",
      "After 3 runs, maximum modularity is Q = 0.606142\n",
      "Louvain completed 23 runs in 17.32654118537903 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 111.31317472457886 seconds\n",
      "epoch: 250, \t total loss: 704.51996,\t reconstruction loss: 628.20703,\t sparse penalty: 0.20882,\t x triplet loss: 7.53952,\t y triplet loss: 7.51423\n",
      "epoch: 300, \t total loss: 704.61743,\t reconstruction loss: 628.47937,\t sparse penalty: 0.19899,\t x triplet loss: 7.52580,\t y triplet loss: 7.50281\n",
      "epoch: 350, \t total loss: 702.89240,\t reconstruction loss: 626.90247,\t sparse penalty: 0.19003,\t x triplet loss: 7.51439,\t y triplet loss: 7.49358\n",
      "epoch: 400, \t total loss: 702.76593,\t reconstruction loss: 626.92102,\t sparse penalty: 0.18276,\t x triplet loss: 7.50440,\t y triplet loss: 7.48182\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 57.99834656715393 seconds\n",
      "Jaccard graph constructed in 25.82192873954773 seconds\n",
      "Wrote graph to binary file in 1.0400006771087646 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.838735\n",
      "After 2 runs, maximum modularity is Q = 0.839799\n",
      "Louvain completed 22 runs in 14.349984169006348 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 120.67132139205933 seconds\n",
      "Finding 30 nearest neighbors using minkowski metric and 'auto' algorithm\n",
      "Neighbors computed in 52.56972813606262 seconds\n",
      "Jaccard graph constructed in 25.993175983428955 seconds\n",
      "Wrote graph to binary file in 1.0361874103546143 seconds\n",
      "Running Louvain modularity optimization\n",
      "After 1 runs, maximum modularity is Q = 0.602826\n",
      "After 4 runs, maximum modularity is Q = 0.607801\n",
      "After 12 runs, maximum modularity is Q = 0.611082\n",
      "Louvain completed 32 runs in 22.41238236427307 seconds\n",
      "Sorting communities by size, please wait ...\n",
      "PhenoGraph complete in 124.90429329872131 seconds\n",
      "epoch: 450, \t total loss: 701.68268,\t reconstruction loss: 625.96307,\t sparse penalty: 0.17759,\t x triplet loss: 7.49240,\t y triplet loss: 7.47394\n",
      "++++++++++ MUSE completed ++++++++++\n"
     ]
    }
   ],
   "source": [
    "## this is just for testing\n",
    "\n",
    "################################# loop batch version, random sample\n",
    "for i in range(5):\n",
    "\n",
    "  np.random.seed(i)\n",
    "  indices = np.random.choice(datab_full.shape[0], 10000, replace=False)\n",
    "  dataa_sub = dataa.iloc[indices,:]\n",
    "  datab_sub = datab_full[indices,:]\n",
    "  view_a_label_sub = view_a_label[indices]\n",
    "  view_b_label_sub = view_b_label[indices]\n",
    "\n",
    "  muse_feature, reconstruct_x, reconstruct_y, \\\n",
    "    latent_x, latent_y = muse_fit_predict(dataa_sub.to_numpy(),\n",
    "                                          datab_sub,\n",
    "                                          view_a_label_sub,\n",
    "                                          view_b_label_sub,\n",
    "                                          latent_dim=30,\n",
    "                                          n_epochs=500,\n",
    "                                          lambda_regul=5,\n",
    "                                          lambda_super=5)\n",
    "  name = 'batch_' + str(i) + '.npy'\n",
    "  np.save('../data/muse_469full_embedding_' + name, muse_feature)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "muse",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
